### ISSUE TITLE
tensorflow/lite/kernels/gather.cc:80 0 <= axis && axis < NumDimensions(input). tflite for android

### ISSUE THREAD

**Issue by bmabir17** [tensorflow/tensorflow#39179]
I have converted a [mask-rcnn model](https://github.com/matterport/Mask_RCNN/releases/tag/v2.0) to deploy it on android. I was able to load the keras weight, freeze the model and convert it to .tflite model using **tflite 1.13 toco** using [this script](https://gist.github.com/bmabir17/754a6e0450ec4fd5e25e462af949cde6). This model contains ResizeNearestNeighbor, Stack, and TensorFlowShape tf_ops. So i had to use 
`converter.target_ops = [tf.lite.OpsSet.TFLITE_BUILTINS,tf.lite.OpsSet.SELECT_TF_OPS]`
Because of this i have already add the tflite build with select-tf-ops like the following
```java
dependencies {
    implementation('org.tensorflow:tensorflow-lite:0.0.0-nightly'){ changing = true }
    implementation('org.tensorflow:tensorflow-lite-gpu:0.0.0-nightly'){ changing = true }
    implementation('org.tensorflow:tensorflow-lite-select-tf-ops:0.0.0-nightly'){ changing = true }
    implementation('org.tensorflow:tensorflow-lite-support:0.0.0-nightly'){ changing = true }
}
```

Now i am trying to use this converted model for inference in android.
```java
    protected Classifier(Activity activity, Device device, int numThreads) throws IOException {
        tfliteModel = FileUtil.loadMappedFile(activity, getModelPath());
        switch (device) {
            case GPU:
                // TODO: Create a GPU delegate instance and add it to the interpreter options

                break;
            case CPU:
                break;
        }
        tfliteOptions.setNumThreads(numThreads);
        tflite = new Interpreter(tfliteModel, tfliteOptions);

        // Loads labels out from the label file.
        labels = FileUtil.loadLabels(activity, getLabelPath());
        // Todo change output as multiple output for maskrcnn
        // Reads type and shape of input and output tensors, respectively.
        int imageTensorIndex = 0;
        int[] imageShape = tflite.getInputTensor(imageTensorIndex).shape(); // {1, height, width, 3}
        imageSizeY = imageShape[1];
        imageSizeX = imageShape[2];
        DataType imageDataType = tflite.getInputTensor(imageTensorIndex).dataType();
        int probabilityTensorIndex = 0;
        int[] probabilityShape =
                tflite.getOutputTensor(probabilityTensorIndex).shape(); // {1, NUM_CLASSES}
        DataType probabilityDataType = tflite.getOutputTensor(probabilityTensorIndex).dataType();
        // Creates the input tensor.
        inputImageBuffer = new TensorImage(imageDataType);

        // Creates the output tensor and its processor.
        outputProbabilityBuffer = TensorBuffer.createFixedSize(probabilityShape, probabilityDataType);

        // Creates the post processor for the output probability.
        probabilityProcessor = new TensorProcessor.Builder().add(getPostprocessNormalizeOp()).build();

        LOGGER.d("Created a Tensorflow Lite Image Classifier.");
    }
    public List<Recognition> InferImage(final Bitmap bitmap) {
        int sensorOrientation=0;
        inputImageBuffer = loadImage(bitmap, sensorOrientation);
        tflite.run(inputImageBuffer.getBuffer(), outputProbabilityBuffer.getBuffer().rewind());

        // Gets the map of label and probability
        Map<String, Float> labeledProbability =
                new TensorLabel(labels, probabilityProcessor.process(outputProbabilityBuffer))
                        .getMapWithFloatValue();
        // Gets top-k results.
        return getTopKProbability(labeledProbability);
    }
```
[Full Source Code](https://gist.github.com/bmabir17/4ba3df476b5e27c38a4979e0e627f275)
### But the line with tflite.run throws the following error
`tensorflow/lite/kernels/gather.cc:80 0 <= axis && axis < NumDimensions(input).`

Detailed Logcat
```
2020-05-05 16:46:05.199 8663-8755/co.chowagiken.tflite_android E/AndroidRuntime: FATAL EXCEPTION: inference
    Process: co.chowagiken.tflite_android, PID: 8663
    java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: tensorflow/lite/kernels/gather.cc:80 0 <= axis && axis < NumDimensions(input) was not true.
    Node number 222 (GATHER) failed to prepare.
    
        at org.tensorflow.lite.NativeInterpreterWrapper.run(Native Method)
        at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:154)
        at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:314)
        at org.tensorflow.lite.Interpreter.run(Interpreter.java:275)
        at co.chowagiken.tflite_android.Classifier.InferImage(Classifier.java:196)
        at co.chowagiken.tflite_android.ClassifierActivity$1.run(ClassifierActivity.java:93)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:201)
        at android.os.HandlerThread.run(HandlerThread.java:65)

```

#### Converted Model
[mask_rcnn_coco_1024_quantize.tflite](https://drive.google.com/file/d/1K12f-NCQy52XmMqvnqMsgEQQA7PTTgRq/view?usp=sharing)

**Comment by bmabir17** [comment]
BTW, i have also tried to run inference on the converted model using the tf_lite python api. But it throws segmentation error on `interpreter.invoke()` and python script crashes. I have tried to find a reason why, most of the issues mention that `tf.lite.OpsSet.SELECT_TF_OPS` is still not supported on python interpreter.
[Detailed Issue discussed on stackoverflow](https://stackoverflow.com/questions/61571375/convert-mask-rcnn-model-for-android-with-tflite)

**Comment by bmabir17** [comment]
@ymodak can you please help to understand why this error is error is being thrown?

**Comment by jdduke** [comment]
`ResizeNearestNeighbor, Stack, and TensorFlowShape` should all be supported in the latest TF 2.2 release. Can you give that a try? That should avoid the need to use the TF ops.

**Comment by bmabir17** [comment]
@jdduke I have tried TF 2.2 But the problem persists.
Is this because of image size mismatch from camera which takes image of 640x360 ?
While my model takes input of 1024? I tried to capture image of 1024*1024. But android does not seems to support that. I am kinda new on android 
```2020-06-03 21:58:37.991 13579-13579/co.chowagiken.tflite_android I/tensorflow: CameraConnectionFragment: Desired size: 256x256, min size: 320x320
2020-06-03 21:58:37.991 13579-13579/co.chowagiken.tflite_android I/tensorflow: CameraConnectionFragment: Valid preview sizes: [4032x3024, 4000x3000, 4032x2268, 4032x2016, 3840x2160, 2880x2156, 2688x1512, 2592x1940, 2592x1458, 2592x1296, 1920x1440, 1920x1080, 1600x1200, 1280x960, 1280x720, 1280x640, 800x600, 720x480, 640x480, 640x360]
2020-06-03 21:58:37.991 13579-13579/co.chowagiken.tflite_android I/tensorflow: CameraConnectionFragment: Rejected preview sizes: [352x288, 320x240, 176x144]
2020-06-03 21:58:37.991 13579-13579/co.chowagiken.tflite_android I/tensorflow: CameraConnectionFragment: Chosen size: 640x360
```

[CameraActivity.java](https://gist.github.com/bmabir17/4ba3df476b5e27c38a4979e0e627f275#file-cameraactivity-java)

**Comment by bmabir17** [comment]
I was going through the converted model and found that tflite converted model has some difference compared with the original model.pb(saved before conversion to tflite).
It seems the range op conversion has some problem. And  this seems to be the input for the expand dims.
_Tensorflow Model(original)_
![tensorflow_range](https://user-images.githubusercontent.com/5344072/85119479-a3454400-b243-11ea-946f-ad6ece115221.png)
_Tflite Model(converted)_
![tflite_range](https://user-images.githubusercontent.com/5344072/85119483-a50f0780-b243-11ea-835d-9fa2c77deba9.png)

**Comment by bmabir17** [comment]
@jdduke 
> `ResizeNearestNeighbor, Stack, and TensorFlowShape` should all be supported in the latest TF 2.2 release. Can you give that a try? That should avoid the need to use the TF ops.

This is [Notebook](https://colab.research.google.com/drive/1JYDYX6X8pSYjsWcTQt87l2Xt0Lzj7JRW?usp=sharing) contains all the different ways i tried to convert the model using tensorflow 2.2 . Only one of the conversion process is successful, and it still requires `tf_ops` to convert successfully.

**Comment by sushreebarsa** [comment]
@bmabir17 It looks like you are using an older Version of Tensorflow . Many bugs have been fixed in the latest version. Could you please execute your code using Latest Version 2.4.1 or 2.5 and let us know if the issue still persists? Thanks!

**Comment by srjoglekar246** [comment]
Closing this bug due to staleness. There was some discussion on [this issue](https://github.com/tensorflow/tensorflow/issues/40297#issuecomment-648988845) about converting the matterport model to TFLite, and looks like the user succeeded. @bmabir17 If this issue still persists (sorry about tghe extremely delayed response), then feel free to reopen.
